### ISSUE TITLE
Shortcut for iterating over lines of a file?

### ISSUE THREAD

**Issue by zevv** [janet-lang/janet#1148]
Janet seems to miss a short and convenient way for the very common use case of iterating over all the lines of a given file; this is something that's very often used for scripting and ad-hoc programs for analyzing or transforming data. I'm looking for something like `io.lines()` in Lua.

Would something like this be a candidate to be added to the `file/` stdlib?

just some implementation below, can probably be improved on:

```
(defmacro file/each-line
  ``Iterate over all the lines of a file; `file` can be a file descriptor or a file name.``
  [line file & body]
  ~(with [fd (if (string? ,file) (file/open ,file) ,file)]
    (forever
      (if (def line (file/read fd :line))
        (do ,;body)
        (break)))))

(file/each-line line "/etc/services"
  (print (length line)))
```

**Comment by sogaiu** [comment]
Sounds nice to me -- though I don't decide things :)

**Comment by bakpakin** [comment]
Here is a functional implementation with a fiber iterator. The macro is a bit more efficient, though.

```janet
(defn file-lines
  "Return an iterator over the lines of a file"
  [file-or-path &opt mode]
  (default mode :r)
  (if (bytes? file-or-path)
    (coro
      (with [f (file/open file-or-path mode)]
        (while (def line (file/read f :line))
          (yield line))))
    (coro
      (while (def line (file/read file-or-path :line))
        (yield line)))))
```

**Comment by sogaiu** [comment]
I like the idea and am curious why it might not be good for something like this to be in the standard library.

**Comment by primo-ppcg** [comment]
Iterating over each line of a file is a common enough task that it may warrant a convenience method. I don't know that it's necessary or desirable for that iterator to also open a file descriptor if passed a string argument, though.

```janet
#currently
(with [myfile (file/open "/etc/services" :r)]
  (while (def line (file/read myfile :line)))
    (prin "Service: " line)))

#proposed
(each line (file/lines "/etc/services")
  (prin "Service: " line))
```

The first of these seems more clear in intent, at least to me, although the `while def` is perhaps a bit ugly. This, less so:

```janet
(with [myfile (file/open "/etc/services" :r)]
  (each line (file/lines myfile))
    (prin "Service: " line)))

**Comment by zevv** [comment]
I think for me the primary use case for this is convenience for ad-hoc scripting, where Janet is a nice step-up from shell scripts; quickly jotting down a few lines to do a task and go on with life.

I think it would benefit from a concise notation and should be easy to work with both files and stdin; note the macro implementation from my original proposal that has an ever simpler notation:

```
(file/each-line line "/etc/services"
  (print (length line)))
```

For example, compare with python:
```
for line in sys.stdin:
    print(line)
```
or Lua:
```
for l in io.lines() do
  print(line
end
```
I'm even old enough to do perl sometimes:
```
while (<>) {
   print;
}
```

**Comment by sogaiu** [comment]
Since in my Perl days I tended to go with a more verbose expression of this intent (e.g. writing out `<STDIN>`, etc.), I'm quite ok with the latter from @primo-ppcg's comparison.  That's brief enough for me :)

**Comment by zevv** [comment]
Closed by 70f13f1b62f02a8d02868d1eea5c9709157de6ca
